Index: include/sdk_events.h
===================================================================
--- include/sdk_events.h	(révision 10118)
+++ include/sdk_events.h	(copie de travail)
@@ -418,6 +418,18 @@
 #define EVT_CLEAN_WORKSPACE_STARTED(fn) DECLARE_EVENT_TABLE_ENTRY( cbEVT_CLEAN_WORKSPACE_STARTED, -1, -1, (wxObjectEventFunction)(wxEventFunction)(CodeBlocksEventFunction)&fn, (wxObject *) NULL ),
 extern EVTIMPORT const wxEventType cbEVT_COMPILER_SETTINGS_CHANGED;
 #define EVT_COMPILER_SETTINGS_CHANGED(fn) DECLARE_EVENT_TABLE_ENTRY( cbEVT_COMPILER_SETTINGS_CHANGED, -1, -1, (wxObjectEventFunction)(wxEventFunction)(CodeBlocksEventFunction)&fn, (wxObject *) NULL ),
+// Defines the futur current state of the compiler.
+// to use with 'cbEVT_COMPILER_STARTED' in  'm_commandInt' or  'm_clientData'
+enum cbFutureBuild
+{
+    fbNone = 0,
+    fbBuild,
+    fbClean,
+    fbRebuild,
+    fbWorkspaceBuild,
+    fbWorkspaceClean,
+    fbWorkspaceReBuild
+};
 
 // request app to compile a single file
 extern EVTIMPORT const wxEventType cbEVT_COMPILE_FILE_REQUEST;
Index: plugins/compilergcc/compilergcc.cpp
===================================================================
--- plugins/compilergcc/compilergcc.cpp	(révision 10118)
+++ plugins/compilergcc/compilergcc.cpp	(copie de travail)
@@ -311,6 +311,7 @@
     m_LastBuildStep(true),
     m_RunTargetPostBuild(false),
     m_RunProjectPostBuild(false),
+    m_Workspace(false),
     m_IsWorkspaceOperation(false),
     m_IsCompileFileRequest(false),
     m_LogBuildProgressPercentage(false)
@@ -352,6 +353,7 @@
     m_Clean = false;
     m_Build = false;
     m_LastBuildStep = true;
+    m_Workspace = false;
     m_IsWorkspaceOperation = false;
     m_IsCompileFileRequest = false;
 
@@ -1040,7 +1042,12 @@
 
     wxString fname = file.GetFullPath();
     if (!fname.IsEmpty())
-        CompileFile( UnixFilename(fname) );
+    {
+        wxString filename = UnixFilename(fname);
+        // fired  cbEVT_COMPILER_STARTED
+        DoPreGenFile(filename);
+        CompileFile(filename);
+    }
 }
 
 wxString CompilerGCC::ProjectMakefile()
@@ -2095,6 +2102,7 @@
     m_pLastBuildingTarget  = 0;
     m_BuildingTargetName   = target;
     m_CommandQueue.Clear();
+    m_Workspace            = job == bjWorkspace;
 }
 
 void CompilerGCC::ResetBuildState()
@@ -2294,6 +2302,8 @@
 
         case bsTargetPreBuild:
         {
+            // fired  cbEVT_COMPILER_STARTED
+            DoPreGenAll();
             // check if it should build with "All"
             // run target pre-build steps
             cmds = dc.GetPreBuildCommands(bt);
@@ -3870,3 +3880,57 @@
     return wxString::Format(_("%d minute(s), %d second(s)"), mins, secs);
 #endif // NO_TRANSLATION
 }
+
+cbFutureBuild CompilerGCC::GetFutureBuild()
+{
+    enum cbFutureBuild FBuild = fbNone;
+    if (! m_Workspace) {
+        if (! m_Clean && m_Build)
+            FBuild = fbBuild;
+        else
+        if (m_Clean && ! m_Build)
+            FBuild = fbClean ;
+        else
+        if (m_Clean && m_Build)
+            FBuild = fbRebuild ;
+    }
+    else  {
+        if (! m_Clean && m_Build)
+            FBuild = fbWorkspaceBuild;
+        else
+        if (m_Clean && ! m_Build)
+            FBuild = fbWorkspaceClean ;
+        else
+        if (m_Clean && m_Build)
+            FBuild = fbWorkspaceReBuild ;
+    }
+
+    return FBuild;
+}
+
+void CompilerGCC::DoPreGenAll()
+{
+    if (m_CommandQueue.GetCount() == 0)
+    {
+        // project/target pairs
+        CodeBlocksEvent evt(cbEVT_COMPILER_STARTED, 0, m_pBuildingProject, 0, this);
+        evt.SetBuildTargetName(m_BuildingTargetName);
+        evt.SetInt((int)GetFutureBuild());
+        evt.SetString(_T(""));
+        Manager::Get()->ProcessEvent(evt);
+    }
+    Manager::Yield();
+}
+
+void CompilerGCC::DoPreGenFile(const wxString& file)
+{
+    if (m_CommandQueue.GetCount() == 0)
+    {
+        CodeBlocksEvent evt(cbEVT_COMPILER_STARTED, 0, m_pProject, 0, this);
+        evt.SetString(file);
+        evt.SetInt(-1);
+        Manager::Get()->ProcessEvent(evt);
+    }
+    Manager::Yield();
+}
+
Index: plugins/compilergcc/compilergcc.h
===================================================================
--- plugins/compilergcc/compilergcc.h	(révision 10118)
+++ plugins/compilergcc/compilergcc.h	(copie de travail)
@@ -181,6 +181,11 @@
         void SaveOptions();
         void LoadOptions();
         void DoRegisterCompilers();
+
+        cbFutureBuild GetFutureBuild();
+        void DoPreGenAll();
+        void DoPreGenFile(const wxString& file);
+
         void DoPrepareQueue(bool clearLog);
         void NotifyCleanProject(const wxString& target);
         void NotifyCleanWorkspace();
@@ -308,7 +313,7 @@
         // to decide if post-build steps should run
         bool m_RunTargetPostBuild;
         bool m_RunProjectPostBuild;
-
+        bool m_Workspace; // true if Build, Rebuild, Clean workspace
         bool m_IsWorkspaceOperation; // true for workspace commands
         bool m_IsCompileFileRequest; // true for plugin compile file requests
 
