Index: include/cbplugin.h
===================================================================
--- include/cbplugin.h	(révision 11246)
+++ include/cbplugin.h	(copie de travail)
@@ -37,7 +37,7 @@
 // this is the plugins SDK version number
 // it will change when the SDK interface breaks
 #define PLUGIN_SDK_VERSION_MAJOR   1
-#define PLUGIN_SDK_VERSION_MINOR   32
+#define PLUGIN_SDK_VERSION_MINOR   33
 #define PLUGIN_SDK_VERSION_RELEASE 0
 
 // class decls
Index: include/sdk_events.h
===================================================================
--- include/sdk_events.h	(révision 11246)
+++ include/sdk_events.h	(copie de travail)
@@ -478,6 +478,19 @@
 
 // settings events
 
+// Defines the futur current state of the compiler.
+// to use with 'cbEVT_COMPILER_STARTED' in  'm_commandInt' or 'm_clientData'
+enum cbFutureBuild
+{
+    fbNone = 10,
+    fbBuild,
+    fbClean,
+    fbRebuild,
+    fbWorkspaceBuild,
+    fbWorkspaceClean,
+    fbWorkspaceReBuild
+};
+
 // event.GetInt() returns value of type cbSettingsType::Type indicating which setting group was changed
 struct cbSettingsType
 {
Index: plugins/compilergcc/compilergcc.cpp
===================================================================
--- plugins/compilergcc/compilergcc.cpp	(révision 11246)
+++ plugins/compilergcc/compilergcc.cpp	(copie de travail)
@@ -303,6 +303,7 @@
     m_LastBuildStep(true),
     m_RunTargetPostBuild(false),
     m_RunProjectPostBuild(false),
+    m_Workspace(false),
     m_IsWorkspaceOperation(false),
     m_LogBuildProgressPercentage(false)
 {
@@ -343,6 +344,7 @@
     m_Clean = false;
     m_Build = false;
     m_LastBuildStep = true;
+    m_Workspace = false;
     m_IsWorkspaceOperation = false;
 
     m_timerIdleWakeUp.SetOwner(this, idTimerPollCompiler);
@@ -1034,6 +1036,10 @@
         CodeBlocksLogEvent evtSwitch(cbEVT_SWITCH_TO_LOG_WINDOW, m_pLog);
         Manager::Get()->ProcessEvent(evtSwitch);
 
+        wxString filename = UnixFilename(fname);
+        // fired  cbEVT_COMPILER_STARTED
+        DoPreGenFile(filename);
+
         CompileFile( UnixFilename(fname) );
     }
 }
@@ -2159,6 +2165,7 @@
     m_pLastBuildingTarget  = 0;
     m_BuildingTargetName   = target;
     m_CommandQueue.Clear();
+    m_Workspace            = job == bjWorkspace;
 }
 
 void CompilerGCC::ResetBuildState()
@@ -2358,6 +2365,9 @@
 
         case bsTargetPreBuild:
         {
+            if (m_Build)
+            // fired  cbEVT_COMPILER_STARTED
+                DoPreGen(GetFutureBuild());
             // check if it should build with "All"
             // run target pre-build steps
             cmds = dc.GetPreBuildCommands(bt);
@@ -3959,3 +3969,59 @@
     return wxString::Format(_("%d minute(s), %d second(s)"), mins, secs);
 #endif // NO_TRANSLATION
 }
+
+
+cbFutureBuild CompilerGCC::GetFutureBuild()
+{
+    enum cbFutureBuild FBuild = fbNone;
+    if (! m_Workspace)
+    {
+        if (! m_Clean && m_Build)
+            FBuild = fbBuild;
+        else
+        if (m_Clean && ! m_Build)
+            FBuild = fbClean ;
+        else
+        if (m_Clean && m_Build)
+            FBuild = fbRebuild ;
+    }
+    else
+    {
+        if (! m_Clean && m_Build)
+            FBuild = fbWorkspaceBuild;
+        else
+        if (m_Clean && ! m_Build)
+            FBuild = fbWorkspaceClean ;
+        else
+        if (m_Clean && m_Build)
+            FBuild = fbWorkspaceReBuild ;
+    }
+
+    return FBuild;
+}
+
+void CompilerGCC::DoPreGen(const cbFutureBuild action)
+{
+    if (m_CommandQueue.GetCount() == 0)
+    {
+        // project/target pairs + cbFutureBuild
+        CodeBlocksEvent evt(cbEVT_COMPILER_STARTED, 1, m_pBuildingProject, 0, this);
+        evt.SetBuildTargetName(m_BuildingTargetName);
+        evt.SetInt(static_cast<int>(action));
+        Manager::Get()->ProcessEvent(evt);
+    }
+    Manager::Yield();
+}
+
+void CompilerGCC::DoPreGenFile(const wxString& file)
+{
+    if (m_CommandQueue.GetCount() == 0)
+    {
+        CodeBlocksEvent evt(cbEVT_COMPILER_STARTED, 1, m_pProject, 0, this);
+        evt.SetString(file);
+        evt.SetInt(static_cast<int>(fbBuild));
+        Manager::Get()->ProcessEvent(evt);
+    }
+    Manager::Yield();
+}
+
Index: plugins/compilergcc/compilergcc.h
===================================================================
--- plugins/compilergcc/compilergcc.h	(révision 11246)
+++ plugins/compilergcc/compilergcc.h	(copie de travail)
@@ -179,6 +179,11 @@
         void SaveOptions();
         void LoadOptions();
         void DoRegisterCompilers();
+
+        cbFutureBuild GetFutureBuild();
+        void DoPreGen(const cbFutureBuild action) ;
+        void DoPreGenFile(const wxString& file);
+
         void DoPrepareQueue(bool clearLog);
         void NotifyCleanProject(const wxString& target);
         void NotifyCleanWorkspace();
@@ -313,7 +318,7 @@
         // to decide if post-build steps should run
         bool m_RunTargetPostBuild;
         bool m_RunProjectPostBuild;
-
+        bool m_Workspace; // true if Build, Rebuild, Clean workspace
         bool m_IsWorkspaceOperation; // true for workspace commands
 
         wxString   m_BuildLogFilename;
